// pyrite/lib/src/component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const decorators_1 = require("./decorators");
class Component {
    constructor(vNode) {
        this.props = vNode.attrs;
        const template = Reflect.getMetadata(decorators_1.templateSymbol, this);
        if (template)
            this.view = render.bind(this, template);
        if (this.props.ref)
            this.props.ref(this);
    }
    $onInit(vNode) { }
    $onCreate(vNode) { }
    $onBeforeUpdate(vNode, oldvNode) { }
    $onUpdate(vNode) { }
    $onBeforeRemove(vNode) { }
    $onRemove(vNode) { }
    oninit(vNode) {
        return this.$onInit(vNode);
    }
    oncreate(vNode) {
        return this.$onCreate(vNode);
    }
    onbeforeupdate(vNode, oldvNode) {
        return this.$onBeforeUpdate(vNode, oldvNode);
    }
    onupdate(vNode) {
        return this.$onUpdate(vNode);
    }
    onbeforeremove(vNode) {
        return this.$onBeforeRemove(vNode);
    }
    onremove(vNode) {
        return this.$onRemove(vNode);
    }
    view(vNode) {
        return null;
    }
}
exports.Component = Component;
function render(template, vNode) {
    this.children = vNode.children;
    if (this.preventDraw) {
        this.preventDraw = false;
        return null;
    }
    if (this.decorators) {
        this.decorators.forEach((decorator) => decorator.call(this));
    }
    return template.call(this, vNode);
}
